src/client/ssl_async.cc

```diff
@@ -0,0 +1,1266 @@
+      // SSL_OP_IGNORE_UNEXPECTED_EOF set
+    } else {
+        #ifdef DEBUG
+        long error =
+        #endif
+            ERR_get_error();
+        #ifdef DEBUG
+        const char* error_string = ERR_error_string(error, NULL);
+        PS_LOG_DEBUG_ARGS("Could not SSL_read (returned <= 0), "
+                          "ssl_error %d, ERR %ld (%s)",
+                          ssl_error,
+                          error, error_string);
+        #endif
+    }
+
+    // ECONNRESET: Connection reset by peer; or else, no data available
+    errno = ((ssl_error == SSL_ERROR_ZERO_RETURN) ? ECONNRESET : ENODATA);
+    return BREAK;
+  } else {
+      mReadFromVec.insert(mReadFromVec.end(), &(buffer[0]), &(buffer[num]));
+
+      PS_LOG_DEBUG_ARGS("read %d bytes", num);
+      /*
+       * Use this one instead to output the whole read buffer
+      PS_LOG_DEBUG_ARGS("read %d bytes: %s", num, &(buffer[0]));
+      */
+  }
+
+  return NEITHER;
+}
+
+// ---------------------------------------------------------------------------
+
+
+SslAsync::ACTION SslAsync::ssl_write()
+{
+  PS_LOG_DEBUG("calling SSL_write");
+
+  if (mCallSslWriteForSslLib && (mToWriteVec.empty())) {
+    PS_LOG_INFO("ssl should not have requested a write from a read if no data was waiting to be written");
+    return BREAK;
+  }
+
+  mCallSslWriteForSslLib = 0;
+
+  if (mToWriteVec.empty())
+    return NEITHER;
+
+  int num = SSL_write(mSsl, mToWriteVec.data(),
+                      static_cast<int>(mToWriteVec.size()));
+  if (num == 0) {
+    long error = ERR_get_error();
+    const char* error_str = ERR_error_string(error, NULL);
+    PS_LOG_INFO_ARGS("could not SSL_write (returned 0): %s", error_str);
+    return BREAK;
+  } else if (num < 0) {
+    int ssl_error = SSL_get_error(mSsl, num);
+    if (ssl_error == SSL_ERROR_WANT_WRITE) {
+      PS_LOG_DEBUG("SSL_write wants write");
+      mWantsTcpWrite = 1;
+      return CONTINUE;
+    }
+
+    if (ssl_error == SSL_ERROR_WANT_READ) {
+      PS_LOG_DEBUG("SSL_write wants read");
+      mCallSslReadForSslLib = 1;
+      // wants_tcp_read is always 1;
+      return CONTINUE;
+    }
+
+#ifdef DEBUG
+    long error =
+#endif
+        ERR_get_error();
+#ifdef DEBUG
+    const char* error_string = ERR_error_string(error, NULL);
+    PS_LOG_DEBUG_ARGS("could not SSL_write (returned -1): %s", error_string);
+#endif
+    return BREAK;
+  } else {
+      PS_LOG_DEBUG_ARGS("wrote %d of %d bytes", num, mToWriteVec.size());
+      if (mToWriteVec.size() < static_cast<std::size_t>(num))
+      {
+          mToWriteVec.clear();
+          mWantsTcpWrite = 1;
+      }
+      else
+      {
+          mToWriteVec.erase(mToWriteVec.begin(), mToWriteVec.begin()+num);
+          mWantsTcpWrite = 0;
+      }
+  }
+
+  return NEITHER;
+}
+
+// ---------------------------------------------------------------------------
+
+
+// Helper function to be called only by initOpenSslIfNotAlready
+static void init_openssl_library(void)
```

Comment:
Wait, is this needed?

Library initialization is a hairy problem, especially for libraries (like pistache).

What if Pistache is embedded into another program, which performs its own OpenSSL initialization?

Reply:
I'm not sure where `init_openssl_library` is defined, but if it's calling `SSL_library_init` then it was [deprecated](https://docs.openssl.org/master/man3/SSL_library_init/#warnings) in OpenSSL 1.1.0 by `OPENSSL_init_ssl`.

Reply:
See below